### PR TITLE
Moving Pgan into the generic image class and improving coverage

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/optimization/test_optimizerlib.py [hunk 1]
```diff
@@ -642,7 +642,7 @@ def test_ngo_split_optimizer(
 @pytest.mark.parametrize(  # type: ignore
     "budget,with_int",
     [
-        (100, True),
+        (115, True),
```
Comment (teytaud) [inline nevergrad/optimization/test_optimizerlib.py:645]: let us fight flakiness

**Comment by teytaud** [comment]
I am not sure of this part in the code:

./nevergrad/benchmark/test_experiments.py:    with tools.set_env(NEVERGRAD_PYTEST=1):
./nevergrad/functions/rl/test_envs.py:import pytest
./nevergrad/functions/rl/test_envs.py:    with pytest.raises(ValueError):
./nevergrad/functions/rl/test_envs.py:    with pytest.raises(ValueError):
./nevergrad/functions/helpers.py:        if "PYTEST_NEVERGRAD" in os.environ:

There is NEVERGRAD_PYTEST and PYTEST_NEVERGRAD, typo or feature ?

**Comment by teytaud** [comment]
@fteytaud a small fix was needed in kerastuning (3 lines...).

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 2]
```diff
@@ -98,7 +99,11 @@ def _ml_parametrization(
             regr.fit(X, y, **fit_additional_params)
             # Predict
             pred_test = regr.predict(X_test)
-            result += mean_squared_error(y_test, pred_test)
+            try:
+                result += mean_squared_error(y_test, pred_test)
+            except ValueError:
```
Comment (teytaud) [inline nevergrad/functions/ml/mlfunctionlib.py:104]: NaN was killing us.

File: nevergrad/benchmark/test_experiments.py [hunk 3]
```diff
@@ -26,15 +28,36 @@
...
+    if "tuning" in name and platform.system() == "Windows":
+        raise SkipTest("Image quality not guaranteed on Windows.")
+
+    # Basic test.
     with tools.set_env(NEVERGRAD_PYTEST=1):
         with datasets.mocked_data():  # mock mlda data that should be downloaded
             check_maker(maker)  # this is to extract the function for reuse if other external packages need it
-    if name not in ["rocket", "images_using_gan", "control_problem"] and not any(
-        x in name for x in ["tuning", "mlda", "realworld", "image_"]
+
```
Comment (teytaud) [inline nevergrad/benchmark/test_experiments.py:45]: we were skipping rocket, images_* image_*, realworld, mlda, tuning, control_problem.

I still skip control_problem.
For others, I try to skip minimally: remove 
- the seed testing for rocket and images_using_gan;
- image quality on Windows;
- gan, mlda, realworld, only on the CI.

File: nevergrad/functions/images/core.py [hunk 4]
```diff
@@ -50,16 +53,46 @@ def __init__(
...
-        array = ng.p.Array(init=128 * np.ones(self.domain_shape), mutable_sigma=True)
-        array.set_mutation(sigma=35)
-        array.set_bounds(lower=0, upper=255.99, method="clipping", full_range_sampling=True)
-        max_size = ng.p.Scalar(lower=1, upper=200).set_integer_casting()
-        array.set_recombination(ng.p.mutation.Crossover(axis=(0, 1), max_size=max_size)).set_name("")  # type: ignore
-        super().__init__(loss(reference=self.data), array)
+        if not with_pgan:
+            array = ng.p.Array(init=128 * np.ones(self.domain_shape), mutable_sigma=True)
+            array.set_mutation(sigma=35)
+            array.set_bounds(lower=0, upper=255.99, method="clipping", full_range_sampling=True)
```
Comment (brozi) [inline nevergrad/functions/images/core.py:59]: The pixels are between 0 and 255 so shouldn't the bound be up to 255 ?

File: nevergrad/functions/images/core.py [hunk 5]
```diff
@@ -50,16 +53,46 @@ def __init__(
...
+                pretrained=True,
+                useGPU=False,
+            )
+            self.domain_shape = (1, 512)  # type: ignore
+            initial_noise = np.random.normal(size=self.domain_shape)
+            array = ng.p.Array(init=initial_noise, mutable_sigma=True)
+            array.set_mutation(sigma=35.0)
+            array.set_recombination(ng.p.mutation.Crossover(axis=(0, 1))).set_name("")
+            self._descriptors.pop("use_gpu", None)
+            super().__init__(self._loss, array)
```
Comment (brozi) [inline nevergrad/functions/images/core.py:77]: shouldn't the with_pgan part be in a different class? Here self.pgan_model will not always be defined so _generate_images will fail if you created the instance with pgan_model = False.

Comment (teytaud) [reply nevergrad/functions/images/core.py:77]: Why would we call _generate_images if there is no pgan_model ?

File: nevergrad/benchmark/experiments.py [hunk 6]
```diff
@@ -1162,11 +1162,11 @@ def arcoating(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
...
 @registry.register
-def image_similarity(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
+def image_similarity(seed: tp.Optional[int] = None, with_pgan: bool = False) -> tp.Iterator[Experiment]:
     """Optimizing images: artificial criterion for now."""
     seedg = create_seed_generator(seed)
     optims = get_optimizers("structured_moo", seed=next(seedg))
-    func = imagesxp.Image()
+    func = imagesxp.Image(with_pgan=with_pgan)
     for budget in [100 * 5 ** k for k in range(3)]:
         for num_workers in [1]:
```
Comment (fteytaud) [inline nevergrad/benchmark/experiments.py:1171]: Maybe the loop on num_workers is not necessary ?

File: nevergrad/benchmark/experiments.py [hunk 7]
```diff
@@ -1194,19 +1200,31 @@ def image_multi_similarity(seed: tp.Optional[int] = None, cross_valid: bool=Fals
                     yield xp
 
 
+@registry.register
+def image_multi_similarity_pgan(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
+    """Counterpart of image_similarity, using PGan as a representation."""
```
Comment (fteytaud) [inline nevergrad/benchmark/experiments.py:1205]: Just a typo : "Counterpart of image_multi_similarity"

File: nevergrad/benchmark/experiments.py [hunk 8]
```diff
@@ -1194,19 +1200,31 @@ def image_multi_similarity(seed: tp.Optional[int] = None, cross_valid: bool=Fals
...
 @registry.register
 def image_multi_similarity_cv(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
     """Counterpart of image_multi_similarity with cross-validation."""
     return image_multi_similarity(seed, cross_valid=True)
 
 
 @registry.register
-def image_quality_proxy(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
+def image_multi_similarity_pgan_cv(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
+    """Counterpart of image_multi_similarity with cross-validation."""
```
Comment (fteytaud) [inline nevergrad/benchmark/experiments.py:1217]: Just another typo : Counterpart of image_multi_similarity with cross-validation, using PGan as a representation.

**Comment by teytaud** [comment]
Non_proxy_function needed.

File: nevergrad/functions/images/core.py (hunk 5 above)
Comment (teytaud) [reply nevergrad/functions/images/core.py]: After offline discussion, I renamed functions.

File: nevergrad/benchmark/experiments.py (hunk 6 above)
Comment (teytaud) [reply nevergrad/benchmark/experiments.py:1171]: Removed

File: nevergrad/benchmark/experiments.py (hunk 8 above)
Comment (teytaud) [reply nevergrad/benchmark/experiments.py:1217]: Done.

**Pull request merged**